WordPress/gutenberg
# Rename Group blocks in the Editor via Modal

## What?

<img width="560" alt="Screen Shot 2023-08-16 at 14 29 09" src="https://github.com/WordPress/gutenberg/assets/444434/5c553eb6-3647-4eef-926f-fe33080dda3e">



Adds the ability to give blocks (currently limited to Group block) a custom name via a `Rename` item in the block's options menu.

This is then displayed in the List View (although that functionality was enabled in a previous PR).

⚠️ Unlike https://github.com/WordPress/gutenberg/pull/42605 you cannot double click to rename blocks in the List View. See below for why this PR doesn't pursue this approach.

Alternative to 
- https://github.com/WordPress/gutenberg/pull/53852
- https://github.com/WordPress/gutenberg/pull/42605

## Why?


Allowing users to distinguish between blocks in the List View is becoming increasingly important as the scope of the Site Editor grows. Given that all blocks are currently labelled by the block name  (e.g. `Group`) it can be difficult to distinguish between them. This is especially important if your Groups represent distinct "sections" of a given page/template.

As you can see [the feature is much requested](https://github.com/WordPress/gutenberg/pull/42605#issuecomment-1475904839).

### Why not pursue the existing PR for renaming within List View?

https://github.com/WordPress/gutenberg/pull/42605 has been around for a while and was my original attempt at this. Unfortunately it was beset by a number of a11y changes relating to editing/renaming _inline_ with the list view.

In the PR several contributors suggested making the it a global feature where rename could be handled via a modal. Returning to the PR it makes sense to land this simpler and more intuitive approach first and then circle back to the "inline" block editing (within the list view) if that is possible/desirable. 

### Why is this limited to the Group block?

The aim is to limit the scope of the change. Once merged it will be possible to follow up to enable this feature for other "container" type blocks.

### Where are you saving this custom name data?

It's saved to the block's `metadata` attribute under the `name` property. That's in WP Core and standard.

```json
{
    "attributes": {
        "metadata": {
            "name": "My custom name"
        }
    }
}
```

## How?


Adds a `Rename` menu item to the block's options menu (available in canvas or in list view). When clicked a modal appears which allows you to provide a custom name for the block.

Once applied the block's metadata is set with a custom `name` attribute which is then reflected everywhere that consumes `useBlockInformation` (that was added previously in another PR).

This means that List View will show the custom name.

## Testing Instructions





- New Post (or enter the Site Editor)
- Add some blocks and then Group them.
- Open List View
- Now open the block options menu (either via List View or via the block in the canvas) and click `Rename`.
- Type a custom name and save.
- See custom name is now shown in List View.
- Bring up the `Rename` UI again, but this time remove all of the text and click out of the input field (triggering `onBlur`).
- See the value reset to the block's original name (e.g. `Group`). See the `Save` button is active again.
- Save the change and verify it updated. 
- Switch to codeview and check the `metadata.name` attribute was reset (it should _not_ be `Group`).

### Testing Instructions for Keyboard


- Do the same as for `Testing instructions` above but use the keyboard.
- Ensure that modal fields are perceivable and focus is placed inside the modal.
- Check that the form is intuitive to use and has accessible roles.
- Check you can submit / cancel the change.
- Key that focus is correctly restored to wherever you "came" from when you triggered "Rename".

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/444434/a36de689-59cc-498f-a6b4-968c3d3dc940

<img width="687" alt="Screen Shot 2023-08-16 at 14 28 25" src="https://github.com/WordPress/gutenberg/assets/444434/2e944c4f-ee66-443c-a61a-79839721d8ef">
